src/network/networkd.c

```diff
@@ -112,6 +113,10 @@ static int run(int argc, char *argv[]) {
         if (r < 0)
                 return log_error_errno(r, "Could not enumerate nexthop: %m");
 
+        r = fw_ctx_new(&m->fw_ctx);
+        if (r < 0)
+                log_warning_errno(r, "Could not initialize firewall, IPMasquerade= option not available: %m");
```

Comment:
Can you clarify this a bit? I can change the api like this:
+int fw_add_masquerade(
+                FirewallContext **ctx,
+                bool add,
+                int af,
+                const union in_addr_union *source,
+                unsigned int source_prefixlen);
+int fw_nftables_add_local_dnat(
+                FirewallContext **ctx,
[..]
(Note '**ctx').
Both functions would check *ctx == NULL and call fw_ctx_new() [which would be an internal helper].
Question is what to do on failure. I think it would make sense to still return the initialized context even on error,
which violates coding guidelines. Otherwise, each future call would try to init/probe the kernel again for the missing feature again, as we can't distinguish NULL (first call) from 'already tried, no working backend available'.
For nspawn this might not be a big deal when the error is changed to be fatal.
But for networkd I'd think exiting if masquerade support isn't available is kinda broken. Any suggestion?
If you think that the extra alloc/probe/free isn't that much of an issue I can change this as outlined above.